homeassistant/components/calendar/__init__.py

```diff
@@ -57,6 +66,36 @@ def get_date(date):
     return dt.as_local(dt.parse_datetime(date['dateTime']))
 
 
+def offset_reached(event, offset_time):
+    """Have we reached the offset time specified in the event title."""
+    if event['start'] is None or offset_time == dt.dt.timedelta():
+        return False
+
+    return event['start'] + offset_time <= \
+        dt.now(event['start'].tzinfo)
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command({
+    vol.Required('type'): WS_EVENTS,
+    vol.Required('entity_id'): cv.entity_id
```

Comment:
Isn't it silly that we work with entities? I wonder if we should be allowed to query the platforms and which calendars they have available in a separate command, then query individual calendars on a platform. 

The calendar component is quite the mess I guess :(

Reply:
Like, we're pushing a whole calendar system to act/look like devices just to get the data out, it's wrong.

Reply:
Yeah, my first thought was why are calendars a binary sensor? Makes sense for them to be a sensor with a list of events. You can use automation to create a binary sensor form there if you really needed it. I see all this code for a calendar panel too but the `async_register_built_in_panel` is commented out? It must have been something that was available before I started. 

I tried to use a combination of system health and media player to figure this out. Media player is what was used as an example in the docs and how the websocket currently gets data https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L856,L882

Reply:
The websocket command is fine in this component, we just shouldn't have a component that puts out weird entities. It makes no sense.

Also this websocket command needs a range.

And how are you dealing with recurring events?

Reply:
added date range and max-results. Nothing is done with recurring events.

Reply:
The question is where are they processed? Is it up to the backend to "extend" them and include an event for each occurrence, or will the backend just include the rule and let the frontend figure it out? 

(I sadly now a little too much about RFC5545 😢)

Reply:
I use the Google calendar and reoccurring events are seen as separate events. Looking through the python caldav library recurring events are a single object though. There's no logic in place for this for the old calendar view so when that was working it was left up to the frontend.